packages/ra-ui-materialui/src/list/datagrid/DatagridConfigurable.tsx

```diff
@@ -0,0 +1,136 @@
+import * as React from 'react';
+import { useResourceContext, usePreference, useStore } from 'ra-core';
+
+import { Configurable } from '../../preferences';
+import { Datagrid, DatagridProps } from './Datagrid';
+import { DatagridEditor } from './DatagridEditor';
+
+/**
+ * A Datagrid that users can customize in configuration mode
+ *
+ * @example
+ * import {
+ *     List,
+ *     DatagridConfigurable,
+ *     TextField,
+ * } from 'react-admin';
+ *
+ * export const PostList = () => (
+ *     <List>
+ *         <DatagridConfigurable>
+ *             <TextField source="id" />
+ *             <TextField source="title" />
+ *             <TextField source="author" />
+ *             <TextField source="year" />
+ *         </DatagridConfigurable>
+ *     </List>
+ * );
+ */
+export const DatagridConfigurable = ({
+    preferenceKey,
+    omit,
+    ...props
+}: DatagridConfigurableProps) => {
+    if (props.optimized) {
+        throw new Error(
+            'DatagridConfigurable does not support the optimized prop'
+        );
+    }
+    const resource = useResourceContext(props);
+    const finalPreferenceKey = preferenceKey || `${resource}.datagrid`;
+
+    const [availableColumns, setAvailableColumns] = useStore<
+        { index: string; source: string; label?: string }[]
+    >(`preferences.${finalPreferenceKey}.availableColumns`, []);
+
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const [_, setOmit] = useStore<string[]>(
+        `preferences.${finalPreferenceKey}.omit`,
+        omit
+    );
+
+    React.useEffect(() => {
+        if (availableColumns.length === 0) {
```

Comment:
This will break if the developer changes the children of the datagrid, and a user has already visited the page.